File: src/windows/AmbientContextMcp.Core.Tests/SensitivityFilterTests.cs
using AmbientContextMcp.Core.Hub;
using AmbientContextMcp.Core.Models;
using Xunit;

namespace AmbientContextMcp.Core.Tests;

public class SensitivityFilterTests
{
    private static readonly IReadOnlyList<PrivacyClassification> Classifications =
    [
        new() { Path = "events.media_session_changed", Sensitivity = "medium", DefaultTransmit = false },
        new() { Path = "events.media_session_changed.title", Sensitivity = "high", DefaultTransmit = false },
        new() { Path = "events.media_session_changed.artist", Sensitivity = "high", DefaultTransmit = false },
    ];

    [Fact]
    public void LookupPayloadFieldSensitivity_returns_exact_match()
    {
        var result = SensitivityScopeFilter.LookupPayloadFieldSensitivity(
            eventName: "media_session_changed",
            payloadKey: "title",
            classifications: Classifications,
            fallbackSensitivity: "medium");

        Assert.Equal("high", result);
    }

    [Fact]
    public void LookupPayloadFieldSensitivity_falls_back_to_event_level()
    {
        var result = SensitivityScopeFilter.LookupPayloadFieldSensitivity(
            eventName: "media_session_changed",
            payloadKey: "source_app",
            classifications: Classifications,
            fallbackSensitivity: "medium");

        Assert.Equal("medium", result);
    }

    [Fact]
    public void ComputePayloadSensitivity_returns_max_high_when_any_field_is_high()
    {
        var payload = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            { "title", "Imagine" },
            { "artist", "John Lennon" },
            { "source_app", "Chrome" }
        };

        var (perKey, max) = SensitivityScopeFilter.ComputePayloadSensitivity(
            eventName: "media_session_changed",
            payload: payload,
            classifications: Classifications,
            eventSensitivity: "medium");

        Assert.Equal("high", perKey["title"]);
        Assert.Equal("high", perKey["artist"]);
        Assert.Equal("medium", perKey["source_app"]);
        Assert.Equal("high", max);
    }

    [Fact]
    public void ComputePayloadSensitivity_returns_event_level_when_all_fields_inherit()
    {
        var payload = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            { "from", "battery" },
            { "to", "ac" }
        };

        var (perKey, max) = SensitivityScopeFilter.ComputePayloadSensitivity(
            eventName: "ac_power_connected",
            payload: payload,
            classifications: Classifications,
            eventSensitivity: "low");

        Assert.Equal("low", perKey["from"]);
        Assert.Equal("low", perKey["to"]);
        Assert.Equal("low", max);
    }

    [Fact]
    public void FilterEventForScope_keeps_event_drops_high_keys_when_scope_is_medium()
    {
        var ev = new LocalContextEvent
        {
            Id = "evt_x",
            Sequence = 1,
            ObservedAt = DateTimeOffset.UtcNow,
            Name = "media_session_changed",
            Value = "Imagine",
            Payload = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            {
                { "title", "Imagine" },
                { "source_app", "Chrome" }
            },
            PayloadSensitivity = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            {
                { "title", "high" },
                { "source_app", "medium" }
            },
            Sensitivity = "medium",
            MaxFieldSensitivity = "high"
        };

        var filtered = SensitivityScopeFilter.FilterEventForScope(ev, ["context.medium:read"]);

        Assert.NotNull(filtered);
        Assert.False(filtered!.Payload.ContainsKey("title"));
        Assert.True(filtered.Payload.ContainsKey("source_app"));
        Assert.Equal("medium", filtered.MaxFieldSensitivity);
        Assert.Equal("medium", filtered.PayloadSensitivity["source_app"]);
    }

    [Fact]
    public void FilterEventForScope_drops_event_when_event_level_exceeds_scope()
    {
        var ev = new LocalContextEvent
        {
            Id = "evt_x",
            Sequence = 1,
            ObservedAt = DateTimeOffset.UtcNow,
            Name = "media_session_changed",
            Sensitivity = "medium",
            MaxFieldSensitivity = "medium",
            Payload = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            {
                { "source_app", "Chrome" }
            },
            PayloadSensitivity = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            {
                { "source_app", "medium" }
            }
        };

        var filtered = SensitivityScopeFilter.FilterEventForScope(ev, ["context.low:read"]);

        Assert.Null(filtered);
    }

    [Fact]
    public void FilterEventForScope_passes_through_when_payload_sensitivity_is_empty()
    {
        // 古い events.jsonl から復元したケース。PayloadSensitivity が空でも event-level でフィルタが効く。
        var ev = new LocalContextEvent
        {
            Id = "evt_x",
            Sequence = 1,
            ObservedAt = DateTimeOffset.UtcNow,
            Name = "ac_power_connected",
            Sensitivity = "low",
            MaxFieldSensitivity = "",
            Payload = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            {
                { "to", "ac" }
            }
        };

        var filtered = SensitivityScopeFilter.FilterEventForScope(ev, ["context.low:read"]);

        Assert.NotNull(filtered);
        Assert.True(filtered!.Payload.ContainsKey("to"));
    }
}

public class HubIngestTests
{
    [Fact]
    public void Ingest_populates_payload_sensitivity_from_classifications()
    {
        var hub = LocalContextHubTestFactory.CreateInMemory();
        var observedAt = DateTimeOffset.UtcNow;

        var snapshot = new AmbientContextSnapshot
        {
            ObservedAt = observedAt,
            PrivacyClassifications =
            [
                new() { Path = "events.media_session_changed", Sensitivity = "medium", DefaultTransmit = false },
                new() { Path = "events.media_session_changed.title", Sensitivity = "high", DefaultTransmit = false }
            ],
            OutboundEvents =
            [
                new()
                {
                    ObservedAt = observedAt,
                    Name = "media_session_changed",
                    Value = "Imagine",
                    Payload = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
                    {
                        { "title", "Imagine" },
                        { "source_app", "Chrome" }
                    },
                    Sensitivity = "medium"
                }
            ]
        };

        hub.Ingest(snapshot);

        var poll = hub.PollEvents(new LocalContextPollRequest
        {
            ClientId = "test",
            Scopes = ["context.high:read"],
            Since = observedAt.AddMinutes(-1)
        });

        var ev = Assert.Single(poll.Events);
        Assert.Equal("high", ev.PayloadSensitivity["title"]);
        Assert.Equal("medium", ev.PayloadSensitivity["source_app"]);
        Assert.Equal("high", ev.MaxFieldSensitivity);
    }

    [Fact]
    public void PollEvents_with_medium_scope_strips_high_payload_keys_but_keeps_event()
    {
        var hub = LocalContextHubTestFactory.CreateInMemory();
        var observedAt = DateTimeOffset.UtcNow;

        hub.Ingest(new AmbientContextSnapshot
        {
            ObservedAt = observedAt,
            PrivacyClassifications =
            [
                new() { Path = "events.media_session_changed", Sensitivity = "medium", DefaultTransmit = false },
                new() { Path = "events.media_session_changed.title", Sensitivity = "high", DefaultTransmit = false }
            ],
            OutboundEvents =
            [
                new()
                {
                    ObservedAt = observedAt,
                    Name = "media_session_changed",
                    Value = "Imagine",
                    Payload = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
                    {
                        { "title", "Imagine" },
                        { "source_app", "Chrome" }
                    },
                    Sensitivity = "medium"
                }
            ]
        });

        var poll = hub.PollEvents(new LocalContextPollRequest
        {
            ClientId = "test",
            Scopes = ["context.medium:read"],
            Since = observedAt.AddMinutes(-1)
        });

        var ev = Assert.Single(poll.Events);
        Assert.False(ev.Payload.ContainsKey("title"));
        Assert.Equal("Chrome", ev.Payload["source_app"]);
        Assert.Equal("medium", ev.MaxFieldSensitivity);
    }
}

File: src/macos/Tests/AmbientContextCoreTests/SensitivityFilterTests.swift
import Foundation
import Testing
@testable import AmbientContextCore

@Suite("SensitivityFilter")
struct SensitivityFilterTests {
    static let classifications = [
        PrivacyClassification(path: "events.media_session_changed", sensitivity: "medium", defaultTransmit: false),
        PrivacyClassification(path: "events.media_session_changed.title", sensitivity: "high", defaultTransmit: false),
        PrivacyClassification(path: "events.media_session_changed.artist", sensitivity: "high", defaultTransmit: false)
    ]

    @Test("LookupPayloadFieldSensitivity_returns_exact_match")
    func lookupPayloadFieldSensitivityReturnsExactMatch() {
        let result = SensitivityScopeFilter.lookupPayloadFieldSensitivity(
            eventName: "media_session_changed",
            payloadKey: "title",
            classifications: Self.classifications,
            fallbackSensitivity: "medium")

        #expect(result == "high")
    }

    @Test("LookupPayloadFieldSensitivity_falls_back_to_event_level")
    func lookupPayloadFieldSensitivityFallsBackToEventLevel() {
        let result = SensitivityScopeFilter.lookupPayloadFieldSensitivity(
            eventName: "media_session_changed",
            payloadKey: "source_app",
            classifications: Self.classifications,
            fallbackSensitivity: "medium")

        #expect(result == "medium")
    }

    @Test("ComputePayloadSensitivity_returns_max_high_when_any_field_is_high")
    func computePayloadSensitivityReturnsMaxHighWhenAnyFieldIsHigh() {
        let payload: CaseInsensitiveDictionary<String> = [
            "title": "Imagine",
            "artist": "John Lennon",
            "source_app": "Chrome"
        ]

        let (perKey, max) = SensitivityScopeFilter.computePayloadSensitivity(
            eventName: "media_session_changed",
            payload: payload,
            classifications: Self.classifications,
            eventSensitivity: "medium")

        #expect(perKey["title"] == "high")
        #expect(perKey["artist"] == "high")
        #expect(perKey["source_app"] == "medium")
        #expect(max == "high")
    }

    @Test("ComputePayloadSensitivity_returns_event_level_when_all_fields_inherit")
    func computePayloadSensitivityReturnsEventLevelWhenAllFieldsInherit() {
        let payload: CaseInsensitiveDictionary<String> = ["from": "battery", "to": "ac"]

        let (perKey, max) = SensitivityScopeFilter.computePayloadSensitivity(
            eventName: "ac_power_connected",
            payload: payload,
            classifications: Self.classifications,
            eventSensitivity: "low")

        #expect(perKey["from"] == "low")
        #expect(perKey["to"] == "low")
        #expect(max == "low")
    }

    @Test("FilterEventForScope_keeps_event_drops_high_keys_when_scope_is_medium")
    func filterEventForScopeKeepsEventDropsHighKeysWhenScopeIsMedium() throws {
        let event = LocalContextEvent(
            id: "evt_x",
            sequence: 1,
            observedAt: Date(),
            name: "media_session_changed",
            value: "Imagine",
            payload: ["title": "Imagine", "source_app": "Chrome"],
            sensitivity: "medium",
            payloadSensitivity: ["title": "high", "source_app": "medium"],
            maxFieldSensitivity: "high")

        let filtered = try #require(
            SensitivityScopeFilter.filterEventForScope(event, scopes: ["context.medium:read"]))

        #expect(!filtered.payload.contains("title"))
        #expect(filtered.payload.contains("source_app"))
        #expect(filtered.maxFieldSensitivity == "medium")
        #expect(filtered.payloadSensitivity["source_app"] == "medium")
    }

    @Test("FilterEventForScope_drops_event_when_event_level_exceeds_scope")
    func filterEventForScopeDropsEventWhenEventLevelExceedsScope() {
        let event = LocalContextEvent(
            id: "evt_x",
            sequence: 1,
            observedAt: Date(),
            name: "media_session_changed",
            payload: ["source_app": "Chrome"],
            sensitivity: "medium",
            payloadSensitivity: ["source_app": "medium"],
            maxFieldSensitivity: "medium")

        #expect(SensitivityScopeFilter.filterEventForScope(event, scopes: ["context.low:read"]) == nil)
    }

    @Test("FilterEventForScope_passes_through_when_payload_sensitivity_is_empty")
    func filterEventForScopePassesThroughWhenPayloadSensitivityIsEmpty() throws {
        // 古い events.jsonl から復元したケース。payloadSensitivity が空でも event-level でフィルタが効く。
        let event = LocalContextEvent(
            id: "evt_x",
            sequence: 1,
            observedAt: Date(),
            name: "ac_power_connected",
            payload: ["to": "ac"],
            sensitivity: "low",
            maxFieldSensitivity: "")

        let filtered = try #require(
            SensitivityScopeFilter.filterEventForScope(event, scopes: ["context.low:read"]))
        #expect(filtered.payload.contains("to"))
    }
}

@Suite("HubIngest")
struct HubIngestTests {
    private static func snapshot(observedAt: Date) -> AmbientContextSnapshot {
        AmbientContextSnapshot(
            observedAt: observedAt,
            outboundEvents: [
                AmbientOutboundEvent(
                    observedAt: observedAt,
                    name: "media_session_changed",
                    value: "Imagine",
                    payload: ["title": "Imagine", "source_app": "Chrome"],
                    sensitivity: "medium")
            ],
            privacyClassifications: [
                PrivacyClassification(path: "events.media_session_changed", sensitivity: "medium", defaultTransmit: false),
                PrivacyClassification(path: "events.media_session_changed.title", sensitivity: "high", defaultTransmit: false)
            ])
    }

    @Test("Ingest_populates_payload_sensitivity_from_classifications")
    func ingestPopulatesPayloadSensitivityFromClassifications() throws {
        let hub = LocalContextHubTestFactory.createInMemory()
        let observedAt = Date()

        hub.ingest(Self.snapshot(observedAt: observedAt))

        let poll = hub.pollEvents(LocalContextPollRequest(
            clientId: "test",
            scopes: ["context.high:read"],
            since: observedAt.addingTimeInterval(-60)))

        #expect(poll.events.count == 1)
        let event = try #require(poll.events.first)
        #expect(event.payloadSensitivity["title"] == "high")
        #expect(event.payloadSensitivity["source_app"] == "medium")
        #expect(event.maxFieldSensitivity == "high")
    }

    @Test("PollEvents_with_medium_scope_strips_high_payload_keys_but_keeps_event")
    func pollEventsWithMediumScopeStripsHighPayloadKeysButKeepsEvent() throws {
        let hub = LocalContextHubTestFactory.createInMemory()
        let observedAt = Date()

        hub.ingest(Self.snapshot(observedAt: observedAt))

        let poll = hub.pollEvents(LocalContextPollRequest(
            clientId: "test",
            scopes: ["context.medium:read"],
            since: observedAt.addingTimeInterval(-60)))

        #expect(poll.events.count == 1)
        let event = try #require(poll.events.first)
        #expect(!event.payload.contains("title"))
        #expect(event.payload["source_app"] == "Chrome")
        #expect(event.maxFieldSensitivity == "medium")
    }
}
